pages/api/retrieval.ts

```diff
@@ -0,0 +1,169 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import isRequestAllowed from "@lib/middleware/httpRequestAllowed";
+import { logMessage } from "@lib/logger";
+import { DynamoDBClient, QueryCommand, UpdateItemCommand } from "@aws-sdk/client-dynamodb";
+import { BearerTokenPayload } from "@lib/types";
+import { getBearerToken } from "@lib/middleware/bearerToken";
+import executeQuery from "@lib/integration/queryManager";
+import dbConnector from "@lib/integration/dbConnector";
+import jwt from "jsonwebtoken";
+
+/**
+ * @description
+ * Request type: GET
+ * USAGE: 
+ * curl http://localhost:3000/api/retrieval?maxRecords=10&formID=1
+   -H "Accept: application/json"
+   -H "Authorization: Bearer {token}"
+ * 
+ * Will fetch form's responses up to the maximum number (maxRecords) that was specified.
+ * It required to have a valid bearer token which must carry a valid formID. When all these conditons are met,
+ * Not only it should return form reponses but also mark those responses as retrieved by updating the attribute Retrieved
+ * from 0 to 1.
+ * @param res 
+ * @param formID 
+ * @param maxRecords 
+ * @returns 
+ */
+async function getFormResponses(
+  res: NextApiResponse,
+  formID: string,
+  maxRecords: number
+): Promise<void> {
+  //A list to store submissionIDs
+  let submissionIDlist: (string | undefined)[] = [];
+  try {
+    //Create dynamodb client
+    const db = new DynamoDBClient({ region: process.env.AWS_REGION ?? "ca-central-1" });
+    //Create form's responses db param
+    //NB: A filter expression is applied after a Query finishes, but before the results are returned.
+    //Therefore, a Query consumes the same amount of read capacity, regardless of whether a filter expression is present.
+    const getItemsDbParams = {
+      TableName: "Vault",
+      IndexName: "retrieved-index",
+      Limit: maxRecords,
+      //Cannot use partitin key or sort key such as formID and Retrieved attribute in keyConditionExpression simultaneously
+      KeyConditionExpression: "Retrieved = :zero",
+      ExpressionAttributeValues: { ":formID": { S: formID }, ":zero": { N: "0" } },
+      //A filter expression cannot contain partition key or sort key attributes.
+      //You need to specify those attributes in the key condition expression, not the filter expression.
+      FilterExpression: "FormID = :formID",
+      ProjectionExpression: "FormID,SubmissionID,FormSubmission,Retrieved",
+    };
+    //Get form's responses for formID
+    const formResponses = await db.send(new QueryCommand(getItemsDbParams));
+    //Collecting items submissionIDs for logging and updating items.
+    submissionIDlist =
+      formResponses?.Items?.map((response) => {
+        return response.SubmissionID.S;
+      }) ?? [];
+    if (formResponses.Count && formResponses.Count > 0) {
+      //Preparing a query to set records's attribute Retrieved to 1.
+      for (const submissionID of submissionIDlist as string[]) {
```

Comment:
I would have probably separated this out in a separate try catch block so its clear exactly where the error happened either when retrieving the rows or updating the retrieved status. At the moment it wouldn't be so clear from the error messages

Reply:
Done!